### PR TITLE
2: Thread perf - markdown

### PR DESCRIPTION
**Pull request by chrisnojima**
Some small changes to the markdown parser for links (which basically looks at every word)

Loading the thread storybook in production mode timing for initial render:

master:
943
677.62 (reload)
616 (reload again)

pr:
790
537 (reload)
489 (reload again)

### CODE REVIEW

File: shared/markdown/parser.pegjs [hunk 1]
```diff
@@ -1,29 +1,28 @@
 {
   // Instead of encoding all the bad cases into a more complicated regexp lets just add some simple code here
   // Note: We aren't trying to be 100% perfect here, just getting something that works pretty good and pretty quickly
-  function flatten (input) {
-    const result = []
-    let strs = []
-
-    function visit(x) {
```
Comment (chrisnojima) [inline shared/markdown/parser.pegjs]: dont need to make this function every time

File: shared/markdown/parser.pegjs [hunk 2]
```diff
@@ -167,26 +166,26 @@ NativeEmoji
 LinkChar
  = !(SpecialChar+ (LineTerminatorSequence)) char:NonBlank { return char }
 
-Link 
+Link
   = url:( [([]* ("http"i "s"i? ":")? (LinkChar+) ) & {
-    let URL = [].concat.apply([], url).join('')
```
Comment (chrisnojima) [inline shared/markdown/parser.pegjs]: kinda hard to understand whats going on here

File: shared/markdown/parser.pegjs [hunk 3]
```diff
@@ -167,26 +166,26 @@ NativeEmoji
...
-    if (URL.length < 4) { // 4 chars is the shortest a URL can be (i.e. t.co)
+    // All words basically go into this function so lets not make it slow
+    const maybeQuote = url[0] || []
+    const maybeProtocol = url[1] || []
+    const maybeLink = [] = url[2] || []
+    if (maybeLink.length < 4) { // 4 chars is the shortest a URL can be (i.e. t.co)
```
Comment (chrisnojima) [inline shared/markdown/parser.pegjs]: we can check this before we merge

File: shared/markdown/parser.pegjs [hunk 4]
```diff
@@ -167,26 +166,26 @@ NativeEmoji
...
       return null
     }
-
-    /* Make sure this is a real TLD */
-    const tldMatch = URL.match(tldExp)
-    if (tldMatch) {
-      const tld = tldMatch[4]
-      if (!tlds.includes(tld)) { // tlds is an array of all valid TLDs
+    const link = maybeLink.join('')
+    if (!tldExp.test(link)) { // includes a valid tld?
```
Comment (chrisnojima) [inline shared/markdown/parser.pegjs]: old way did a weak regexp then did a search of the list, instead use a stronger regexp

File: shared/reducers/chat2.js [hunk 5]
```diff
@@ -577,7 +577,10 @@ const rootReducer = (state: Types.State = initialState, action: Chat2Gen.Actions
 
       return state.withMutations(s => {
         s.set('messageMap', messageMap)
-        s.set('messageOrdinals', messageOrdinals)
+        // only if different
+        if (!state.messageOrdinals.equals(messageOrdinals)) {
```
Comment (chrisnojima) [inline shared/reducers/chat2.js]: don't thrash if the list is the same

**Comment by chrisnojima** [comment]
@keybase/react-hackers

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

File: shared/markdown/parser.pegjs [hunk 6]
```diff
@@ -167,26 +166,26 @@ NativeEmoji
...
 
-Link 
+Link
   = url:( [([]* ("http"i "s"i? ":")? (LinkChar+) ) & {
-    let URL = [].concat.apply([], url).join('')
-    if (URL.length < 4) { // 4 chars is the shortest a URL can be (i.e. t.co)
+    // All words basically go into this function so lets not make it slow
+    const maybeQuote = url[0] || []
+    const maybeProtocol = url[1] || []
+    const maybeLink = [] = url[2] || []
```
Comment (akalin-keybase) [inline shared/markdown/parser.pegjs]: intended to be `const maybeLink = url[2] || []`?

Comment (chrisnojima) [reply shared/markdown/parser.pegjs]: fixed, tnx

**Pull request merged**